src/components/RouteLeavingGuard.tsx

```diff
@@ -0,0 +1,51 @@
+import { Location } from 'history';
+import React, { useEffect, useState, FC } from 'react';
+import { Prompt, useHistory } from 'react-router-dom';
+import ConfirmDialog from './ConfirmDialog';
+
+interface RouteLeavingGuardProps {
+  modalDescription: string;
+  modalHeading: string;
+  shouldBlockNavigation: boolean;
+}
+const RouteLeavingGuard: FC<RouteLeavingGuardProps> = ({ modalDescription, modalHeading, shouldBlockNavigation }) => {
+  const [showModal, setShowModal] = useState(false);
+  const [nextLocation, setNextLocation] = useState<Location | null>(null);
+  const [confirmedNavigation, setConfirmedNavigation] = useState(false);
+  const history = useHistory();
+
+  const handleBlockedNavigation = (nextLocation: Location): boolean => {
+    if (!confirmedNavigation && shouldBlockNavigation) {
+      setShowModal(true);
+      setNextLocation(nextLocation);
+      return false;
+    }
+    return true;
+  };
+
+  const handleConfirmNavigationClick = () => {
+    setShowModal(false);
+    setConfirmedNavigation(true);
```

Comment:
kunne dette blitt forenklet ved å legge bare `history.push(nextLocation.pathname)` i denne metoden, samt fjerne useEffect?

Reply:
Nei, det fungerer ikke. Siden setState er async, så blir ikke disse satt med en gang. Så hvis vi flytter `history.push` hit, så må vi trykke på "Ja" to ganger før det navigeres bort. Vi trenger de oppdaterte verdiene av `showModal` og `confirmedNavigation`, og å kjøre `history.push` etter det.